src/include/migraphx/iterator.hpp

```diff
@@ -213,6 +234,16 @@ struct iterator_operators
     }
 };
 
+template <class T, class Category>
+struct iterator_types
+{
+    using reference         = T;
+    using iterator_category = Category;
+    using difference_type   = std::ptrdiff_t;
+    using value_type        = std::decay_t<T>;
+    using pointer           = std::add_pointer_t<std::remove_reference_t<reference>>;
```

Comment:
I get what you're doing here but this looks like of funny at glance. So we always define reference but we just use that for a pointer, which means that removes checks from null? Why not just return reference and not worry about the remove_reference then or is the concern stale pointer when reference goes invalid?